src/main/scala/calc/TreeType.scala

```diff
@@ -0,0 +1,26 @@
+package calc
+
+/**
+  * Types used to annotate the AST.
+  */
+sealed trait TreeType {
+  def isFunction: Boolean = false
+  def isNumber: Boolean = true
+  def typeSuff: String
+}
+
+case object DoubleType extends TreeType {
+  override def typeSuff = "D"
+}
+
+// in our language every parameter is a Double, so we only
+// need to know function's arity
+case class FunctionType(arity: Int) extends TreeType {
+  override def isFunction: Boolean = true
+  override def isNumber: Boolean = false
+  override def typeSuff = "F"
+}
+
+case object NoType extends TreeType {
```

Comment:
Actually, I don't think there's any reason for `NoType` to exist in the compiler for this language.